coderd/database/databasefake/databasefake.go

```diff
@@ -129,6 +131,66 @@ type data struct {
 	lastLicenseID   int32
 }
 
+func validateDatabaseTypeWithValid(v reflect.Value) (handled bool, err error) {
+	if v.CanInterface() {
+		if !strings.Contains(v.Type().PkgPath(), "coderd/database") {
+			return true, nil
+		}
+		// Note: database.Null* types don't have a Valid method, so
+		// let's always skip them just in case.
+		if strings.HasPrefix(v.Type().Name(), "Null") {
+			return true, nil
+		}
+		if valid, ok := v.Interface().(interface{ Valid() bool }); ok {
+			if !valid.Valid() {
+				return true, xerrors.Errorf("invalid %s: %q", v.Type().Name(), v.Interface())
+			}
+		}
+		return true, nil
+	}
+	return false, nil
+}
+
+// validateDatabaseType uses reflect to check if struct properties are types
+// with a Valid() bool fuction set. If so, call it and return an error
+// if false.
+//
+// Note that we only check immediate values and struct fields. We do not
+// recurse into nested structs.
+func validateDatabaseType(args interface{}) error {
+	v := reflect.ValueOf(args)
+
+	// Note: database.Null* types don't have a Valid method, we skip them here
+	// because their embedded types may have a Valid method and we don't want
+	// to bother with checking both that tha Valid field is true and that the
+	// type it embeds validates to true. We would need to check:
+	//
+	//	dbNullEnum.Valid && dbNullEnum.Enum.Valid()
+	if strings.HasPrefix(v.Type().Name(), "Null") {
+		return nil
+	}
```

Comment:
The above comment on the function says we do not go into nested structs. Isn't the `Nullable` case a nested struct?

Reply:
So this is to do nothing in the following scenario:

```go
validateDatabaseType(database.NullThing{}) // HasPrefix Null
```

If we have:

```go
validateDatabaseType(database.MyType{Thing: database.NullThing{}})
```

We ultimately will not validate Thing since we only go one level deep and the validator function doesn't check structs.

This was just to keep the logic (simpler) and not having to look at both `database.NullThing.Valid == true && database.NullThing.Thing.Valid()`.